### PR TITLE
Elixir VNode

### PR DESCRIPTION
**Pull request by fkrause98**
## Changes:
- Working VNode implementation in Elixir.
- Change riax.ex to use the implemented VNode instead of using the Riak's Tutorial VNodes via rpc.
- Change application.ex to start a local node.
- Added VNode Master implementation via riax_sup.ex file (check [here](https://efcasado.github.io/riak-core_intro) why its needed)
- Deleted node{1,2,3}.exs files as they were redundant having the dev{1,2,3}.exs configuration for each node.
- The coverage FSM is still missing, maybe it should be done on another PR.
## How to test:
- Start the nodes in separate terminals with `make node1`,`make node2`, and `make node3`.
- Try to join them with `:riak_core.join('dev1@127.0.0.1')` when in node2's or node3's terminal.
- Use `Riax.put(key, value)` to test it.
- Use `Riax.get(key)` to retrieve the value.
- Use `Riax.ring_status` to see if the handoff is working

### CODE REVIEW

File: lib/riax.ex [hunk 1]
```diff
@@ -1,32 +1,85 @@
...
+  @doc """
+  Store a value tied to a key
+  """
   def put(key, value) do
-    :rpc.call(node(), :rc_example, :put, [key, value])
+    sync_command(key, {:put, {key, value}})
   end
 
+  @doc """
+  Retrive a key's value
```
Comment (Loco1982) [inline lib/riax.ex:10]:
```suggestion
  Retrieve a key's value
```

File: lib/riax.ex [hunk 2]
```diff
@@ -1,32 +1,85 @@
...
+
+  def ping(key) do
+    sync_command(key, {:ping, key})
+  end
+
+  @doc """
+  Return the node's name which is most likely
+  to receive the given key as a parameter
+  """
+  def prefered_node_name(key) do
```
Comment (Loco1982) [inline lib/riax.ex:46]:
```suggestion
  def preferred_node_name(key) do
```

File: lib/riax.ex [hunk 3]
```diff
@@ -1,32 +1,85 @@
...
+    {:ok, {_, name}} = prefered_node(key)
```
Comment (Loco1982) [inline lib/riax.ex:47]:
```suggestion
    {:ok, {_, name}} = preferred_node(key)
```

File: lib/riax.ex [hunk 4]
```diff
@@ -1,32 +1,85 @@
...
+    name
+  end
+
+  @doc """
+  Return the node's {index, name} tuple which is most likely
+  to receive the given key as a parameter
+  """
+  defp prefered_node(key) do
```
Comment (Loco1982) [inline lib/riax.ex:55]:
```suggestion
  defp preferred_node(key) do
```

File: lib/riax.ex [hunk 5]
```diff
@@ -1,32 +1,85 @@
...
   end
 
-  defp find_key(key) do
-    {:ok, keys} = keys()
+  @doc """
+  Use the VNode master to send a command
+  to the VNode that receive the key
+  """
+  defp sync_command(key, command) do
+    {:ok, node} = prefered_node(key)
```
Comment (Loco1982) [inline lib/riax.ex:70]:
```suggestion
    {:ok, node} = preferred_node(key)
```

File: lib/riax_vnode.ex [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+  end
+
+  def delete(state) do
+    Logger.debug("deleting the vnode data")
+    {:ok, %{state | data: %{}}}
+  end
+
+  def handle_handoff_data(bin_data, state = %{data: data}) do
+    {key, val} = :erlang.binary_to_term(bin_data)
+    Logger.debug("receieved handoff data with key-value #{key} #{val}")
```
Comment (Loco1982) [inline lib/riax_vnode.ex:95]:
```suggestion
    Logger.debug("received handoff data with key-value #{key} #{val}")
```

File: lib/riax_vnode.ex [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+  end
+
+  def handle_coverage(:values, _key_spaces, {_, req_id, _}, state = %{data: data}) do
+    Logger.info("Received values coverage: #{state}")
+    values = Map.values(data)
+    {:reply, {req_id, values}, state}
+  end
+
+  def handle_coverage(:clear, _key_spaces, {_, req_id, _}, state) do
+    Logger.info("Receieved clear coverage: #{state} ")
```
Comment (Loco1982) [inline lib/riax_vnode.ex:118]:
```suggestion
    Logger.info("Received clear coverage: #{state} ")
```

**Review by Loco1982 — CHANGES_REQUESTED** [review]
Minor spelling mistakes

**Comment by fkrause98** [comment]
> Minor spelling mistakes

Thanks for pointing it out, I've fixed them.

**Review by facundoolano — APPROVED** [review]

**Pull request merged**